daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/JDOMInfosetOutputter.scala

```diff
@@ -48,14 +48,57 @@ class JDOMInfosetOutputter extends InfosetOutputter
     true
   }
 
-  def startSimple(diSimple: DISimple): Boolean = {
+  def startSimple(diSimple: DISimple, xmlOutputStyle: String): Boolean = {
 
     val elem = createElement(diSimple)
+    val correctFormat = new StringBuilder("");
+    val cdataIntro = "<![CDATA["
+    val cdataOutro = "]]>"
+    var charEntMode: Boolean = false
 
     if (diSimple.hasValue) {
       val text =
         if (diSimple.erd.optPrimType.get.isInstanceOf[NodeInfo.String.Kind]) {
-          remapped(diSimple.dataValueAsString)
+          val s = remapped(diSimple.dataValueAsString)
+
+            if(xmlOutputStyle == "prettyPrintSafe"){
+              val newData = s.replaceAll(">","&gt;").replace("\\r\\n", "&#xE00D;")
+              val readyForCDATA = newData.replaceAll("0x","&#x")
+              //Figure out what mode you have to be in
+              if(readyForCDATA(0) == '&') {
+                charEntMode = true
+              } else {
+                charEntMode = false
+                correctFormat.append(cdataIntro)
+              }
+
+              //Traverse over the string surrounding correct areas with CDATA info
+              for(c <- readyForCDATA) {
+                if(charEntMode) {
+                  correctFormat.append(c)
+                  if(c == ';'){
+                    correctFormat.append(cdataIntro)
+                    charEntMode = false
+                  }
+                } else {
+                  if(c == '&'){
+                    correctFormat.append(cdataOutro)
+                    charEntMode = true
+                  }
+                  correctFormat.append(c)
+                }
+              }
+
+              //You are done with the string. If you are still a non
+	            //char ent then close and finish up.
+              if(!charEntMode){
+                correctFormat.append(cdataOutro)
+              }
+
```

Comment:
There's a lot of special logic and edge cases here needed to convert a string for CDATA. Mike found some potential edge cases we are missing.

I wonder if we can use one of the XML library dependencies we have to perform this logic. I would be surprised if one of them doesn't already have logic for something like this. Let's not reinvent the wheel if possible.

Reply:
Looking at Scala XML for inspiration, this this their logic to convert pretty print to CDATA:

https://github.com/scala/scala-xml/blob/main/shared/src/main/scala/scala/xml/PCData.scala#L23-L34

So essentially they just have a special case to deal with strings that contain the CDATA closing characters (i.e. `]]>`), and everything else is just output exactly as is. Isn't that what we want to do or is it more complicated than that?

Reply:
Not quite. We also need to handle if the data string contains any entity reference or character entities. 

"something &amp; else" needs to become "<![CDATA[something ]]>&amp;<![CDATA[ else]]>"

Reply:
In this example, is `&amp;` actually in the data, or is the data actually `something & else`. In either case though, aren't these both valid:

```xml
<![CDATA[something & else]]>
<![CDATA[something &amp; else]]>
```
Aren't things like entities the xml tags (e.g. `<foo>`) not interpreted in CDATA blocks? So this is valid too:

```xml
<![CDATA[blah <foo> bar <bazr>]]>
```
Doesn't CDATA tell the parser to ignore special characters and just treat them as normal chars until it finds the closing `]]>`?

Seems like we want to take exactly what came out of the Daffodil parse, not do any remapping of entities, and just wrap it in the CDATA brackets?

Reply:
Ah... Yes I think you are correct. The string coming from the DFDL Infoset has not had any characters converted into entities. So if it contains ```&``` then it's a literal character, not some entity thing. We have remapped CR to U+E00D at this point, not ```&#xD;``` , so yeah, the algorithm that is in that PCData.scala code:

```"<![CDATA[%s]]>".format(data.replaceAll("]]>", "]]]]><![CDATA[>"))```

Does seem right.

Reply:
I forgot about the remapping stuff, but I think you're right. Looking some more at the XMLTextInfosetOutputter, it has this line to write strings:
```scala
writer.write(scala.xml.Utility.escape(remapped(simpleVal)))
```
So for this CDATA stuff I think we just don't to the scala.xml.Utility.escape thing, and instead wrap the remapped data in a CDATA block using that line. So maybe this all boils down to a few lines like this?
```scala
val data = remapped(simpleVal)
val escaped = xmlEscapeStyle match {
  XmlEscapeStyle.ENTITIES => scala.xml.Utility.escape(data)
  XmlEscapeStyle.CDATA => "<![CDATA[%s]]>".format(data.replaceAll("]]>", "]]]]><![CDATA[>"))
}
writer.write(escaped)
```